### DISCUSSION TITLE
Improve GraphQL subscriptions and Add filter in subscriptions

### DISCUSSION THREAD

**Discussion by jude-urgent2k** [directus/directus#19713] [Feature Requests]
### Summary

Currently in directus we can make subscriptions and receive data according to different events, but by default when we subscribe we do not receive any data, this feature would therefore include returning the existing data first and these data should take into account the conditions (limit, order by, ...etc). So for summary it would be necessary to return first the existing data and then it would be necessary that in subscriptions we can have the same filters as in a query.

### Basic Example

I’ll just take a GraphQL query here and it should be possible to return me first the initial data and then this subscription should work as a GraphQL query with all the filters.

```
subscription Events_mutated {
    events_mutated(event: create filter: {title: {_contains: "ab"}, start_date: {_gte: "01-01-2024"}}
        offset: 0
        limit: 5 ) {
        key
        event
        data {
            event_id
            deleted
            title
            thumbnail {
                id
             }
            category {
                name
            }
            start_date start_time  end_time end_date
        }
    }
}
```
and this should first return me a list of events and to each new data the data that meets the conditions mentioned in the filters must always be returned.

### Motivation

The motivation comes from the use of other technologies like **Hasura** that made me know this feature. As a use case let’s imagine a mobile application, it needs to receive a list of data in real time, this feature would be interesting because we could include the filters and conditions in the subscription and we could also receive the list of data that would be updated each time a data would be added for example, On top of that, I think it would be easier to manage the display of the data because without this feature it would be a bit difficult: for example if we want to recover the 10 most recently created elements we should first retrieve the data with just a query? then connect to a subscription to receive the new element created and then manually update the data list? would have been a bit too complicated, so this is the kind of problem that this feature would solve again.

### Detailed Design

Subcriptions will always work as they work today, no change will be made at this level, only change will be that it would be possible (and optional) to add filters to the subscription graphql as will be done for any simple query; Another change is that by default the subsciptions will return data that respect the specified contions through filters (if filters are present in the query).

### Must Have:
- WebSockets for Graphql
- The ability to include all the filters found in a simple GraphQL query
- The ability to receive initial data after subscribing as if querying a query
- The possibility that data is received in real time after an event has been triggered (creation, update) taking into account all filters

### Drawbacks

No drawbacks what I can see.

### Alternatives

None.

### Adoption Strategy

Given that filters will be optional, I think this will be appreciated by developers and I do not think that will be a breaking change.I don’t think it will affect any other Directus project because it just adds other possibilities by dealing with the basic operation. Yes we can write a migration.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
